### PR TITLE
EA-127 | Support for Interpretation and status field in EncounterTran…

### PR DESCRIPTION
**Pull request by shashikanthgadgay**
…saction.Observation

### CODE REVIEW

**Comment by shashikanthgadgay** [comment]
@dkayiwa Can you have a look at this PR and merge?

File: api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObservationMapper1_12.java [hunk 1]
```diff
@@ -36,6 +39,29 @@ public  ObservationMapper1_12(ConceptMapper conceptMapper, DrugMapper drugMapper
...
+            Enum obsInterpretation = (Enum) getInterpretation.invoke(obs);
+            Enum obsStatus = (Enum) getStatus.invoke(obs);
+
+            String interpretation = (obsInterpretation != null) ? obsInterpretation.name() : null;
+            String status = (obsStatus != null) ? obsStatus.name() : null;
+
+            observation.setInterpretation(interpretation);
+            observation.setStatus(status);
+        } catch (IllegalAccessException e) {
+            e.printStackTrace();
```
Comment (dkayiwa) [inline api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObservationMapper1_12.java]: Did you get a chance to look at this? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotprintout 😊

File: api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java [hunk 2]
```diff
@@ -26,6 +29,43 @@ public ObsMapper1_12(ConceptService conceptService,
...
+        super.mapObservationProperties(observationData, observation);
+        setInterpretationAndStatus(observation, observationData);
+    }
+
+    private void setInterpretationAndStatus(Obs obs,EncounterTransaction.Observation observationData){
+        try {
+            Class<Enum> interpretationClass = (Class<Enum>) Class.forName("org.openmrs.Obs$Interpretation");
+            Class<Enum> statusClass = (Class<Enum>) Class.forName("org.openmrs.Obs$Status");
+
+            Method setInterpretation =
```
Comment (dkayiwa) [inline api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java]: Can we also add a comment explaining why we have decided to use reflection here?
That way, code maintainers will get some useful background and know that once the module starts using a platform 2.1 dependency, then no more need for this. 😊

File: api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java [hunk 3]
```diff
@@ -26,6 +29,43 @@ public ObsMapper1_12(ConceptService conceptService,
...
+            setInterpretation.invoke(obs, interpretation);
+
+            if (observationData.getStatus() != null) {
+                Enum status = Enum.valueOf(statusClass, observationData.getStatus());
+                setStatus.invoke(obs, status);
+            }
+
+        } catch (NoSuchMethodException e) {
+            e.printStackTrace();
+        } catch (IllegalAccessException e) {
```
Comment (dkayiwa) [inline api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotprintout

File: api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObservationMapper1_12.java [hunk 4]
```diff
@@ -36,6 +39,29 @@ public  ObservationMapper1_12(ConceptMapper conceptMapper, DrugMapper drugMapper
...
         observation.setFormNamespace(obs.getFormFieldNamespace());
         observation.setFormFieldPath(obs.getFormFieldPath());
+        setInterpretationAndStatus(observation, obs);
         return observation;
     }
+
+    private void setInterpretationAndStatus(EncounterTransaction.Observation observation, Obs obs) {
+        try {
+            Method getInterpretation = obs.getClass().getDeclaredMethod("getInterpretation");
+            Method getStatus = obs.getClass().getDeclaredMethod("getStatus");
```
Comment (dkayiwa) [inline api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObservationMapper1_12.java]: Same as above in regards to a comment explaining why we decided to use reflection.

File: api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java (hunk 2 above)
Comment (shashikanthgadgay) [inline api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java]: Sure.. I'll do that.

File: api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java (hunk 3 above)
Comment (shashikanthgadgay) [inline api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java]: Sure.. I'll change it.

**Comment by shashikanthgadgay** [comment]
@dkayiwa Made the changes. Can you have a look and merge it.

File: api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java [hunk 5]
```diff
@@ -26,6 +33,49 @@ public ObsMapper1_12(ConceptService conceptService,
...
+    protected void mapObservationProperties(EncounterTransaction.Observation observationData, Obs observation) {
+        super.mapObservationProperties(observationData, observation);
+        setInterpretationAndStatus(observation, observationData);
+    }
+
+
+    /*
+       This method uses java reflection to set Interpretation and Status as adding a new module dependency
+       for just setting two fields, a lot of duplicate code had to be added.
+       This method can be moved to platform 2.1 dependency, once the platform 2.1 dependency is added.
```
Comment (dkayiwa) [inline api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java]: What do you mean by "This method can be moved to platform 2.1"?

Comment (shashikanthgadgay) [inline api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java]: I meant 
`observation.setInterpretation(interpretation);
 observation.setStatus(status);`
needs to be added to 2.1 platform

Comment (dkayiwa) [inline api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java]: Isn't it already in platform 2.1?

Comment (shashikanthgadgay) [inline api-1.12/src/main/java/org/openmrs/module/emrapi/encounter/mapper/ObsMapper1_12.java]: It is, but if we create a new maven project for 2.1 in emrapi, then we can move it there.

**Comment by shashikanthgadgay** [comment]
@dkayiwa Thanks  :)

**Pull request merged**